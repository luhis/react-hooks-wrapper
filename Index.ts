export { default as stateWrapper } from "./StateWrapper";
export { default as effectWrapper } from "./EffectWrapper";
