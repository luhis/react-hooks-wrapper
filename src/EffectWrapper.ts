import { FunctionComponent, useEffect } from "react";

type EffectResponse = void | (() => void);

type TypeDef = <TProps extends object>
    (effectFuc: (props: TProps) => EffectResponse) =>
    (component: FunctionComponent<TProps>) => FunctionComponent<TProps>;

const EffectWrapper: TypeDef =
    (effectFuc) =>
        (component) =>
            (props) => {
                useEffect(() => effectFuc(props));
                return component(props);
            };

export default EffectWrapper;
