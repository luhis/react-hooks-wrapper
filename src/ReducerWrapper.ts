import { createElement, FunctionComponent, ReactNode, Reducer, useReducer } from "react";

import { Omit, ReconstituteProps, TupleToObject } from "./TypeFunctions";

export default
    <TState extends object, TTransformed extends object, TAction extends object>
        (mapTupleToProps: TupleToObject<TState, TAction, TTransformed>,
         reducer: Reducer<TState, TAction>,
         initialState: TState,
         initialAction?: TAction) =>
        <P extends TTransformed>(component: FunctionComponent<P>) => {
            const ReducerWrapper =  (props: Omit<P, keyof TTransformed> & { readonly children?: ReactNode }) => {
                const transformed = mapTupleToProps(useReducer(reducer, initialState, initialAction));
                const finalProps = ReconstituteProps(props, transformed);
                return createElement(component, finalProps);
            };

            return ReducerWrapper;
        };
