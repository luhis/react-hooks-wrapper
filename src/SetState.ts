import { Dispatch, SetStateAction } from "react";

type SetState<T> = Dispatch<SetStateAction<T>>;

export default SetState;
