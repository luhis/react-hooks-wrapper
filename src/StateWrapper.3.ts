import { createElement, FunctionComponent, ReactNode } from "react";

export default
        <P>(component: FunctionComponent<P>) =>
            (props: P & { readonly children?: ReactNode }) => {
                return createElement(component, props);
            };
