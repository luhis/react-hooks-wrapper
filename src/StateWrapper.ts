import { useState, SFC, ReactElement, Dispatch, SetStateAction } from 'react'

const StateWrapper =
    <TProps extends {}, TState extends {}, TTransformed extends {}>
        (defaultVal: TState, mapTupleToProps: ((tup: [TState, Dispatch<SetStateAction<TState>>]) => TTransformed)) =>
        (component: SFC<TProps & TTransformed>) =>
            (props: TProps): ReactElement<TProps & TTransformed> =>
                component({ ...props, ...mapTupleToProps(useState(defaultVal)) });

export default StateWrapper