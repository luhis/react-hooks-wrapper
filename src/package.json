{
  "name": "hookly",
  "version": "0.0.2",
  "description": "A React plugin to present hooks as higher order functions",
  "keywords": [
    "react",
    "higher-order",
    "components",
    "microcomponentization",
    "toolkit",
    "utilities",
    "composition"
  ],
  "author": "Matt McCorry <mccorry@gmail.com>",
  "license": "ISC",
  "main": "dist/Index.js",
  "types": "dist/Index.d.ts",
  "dependencies": {
  },
  "peerDependencies": {
    "react": "^16.7.0-alpha.2"
  },
  "scripts": {
    "build": "tsc",
    "test": "mocha --reporter spec",
    "lint": "tslint -p tsconfig.json --fix"
  },
  "devDependencies": {
    "@types/react": "^16.7.17",
    "chai": "^4.2.0",
    "mocha": "^5.2.0",
    "react": "^16.7.0-alpha.2",
    "tslint": "^5.12.0",
    "typescript": "^3.2.2"
  },
  "files": ["dist/**/*"]
}
