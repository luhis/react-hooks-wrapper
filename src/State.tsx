import * as React from 'react';

import StateWrapper from './StateWrapper';

type Props = { count: number, setCount: (a: number) => null, name: string };

const Counter: React.SFC<Props> = ({ count, setCount, name }) =>
  <div>
    <p>Hi {name}, You clicked {count} times</p>
    <button onClick={() => setCount(count + 1)}>
      Click me
      </button>
  </div>;

const mapTuple = ([count, setCount]: [number, React.Dispatch<React.SetStateAction<number>>]) => ({ count, setCount });

export default StateWrapper(1, mapTuple)(Counter);