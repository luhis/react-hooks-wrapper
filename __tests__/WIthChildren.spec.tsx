import * as React from "react";
import * as renderer from "react-test-renderer";

import { setState, stateWrapper } from "../src/Index";

interface IFinalComponentProps { readonly name: string; readonly children?: React.ReactNode; }
type Props = IFinalComponentProps & { readonly count: number; readonly setCount: setState<number> };

const Counter: React.FunctionComponent<Props> = ({ count, setCount, name, children }) =>
    <div>
        <p>Hi {name}, You clicked {count} times</p>
        <button onClick={() => setCount(count + 1)}>
            Click me
      </button>
      {children}
    </div>;

const CounterContainer: React.FunctionComponent<IFinalComponentProps>
    = stateWrapper(1, ([count, setCount]) => ({ count, setCount }))(Counter);

test("WithChildren should", () => {
    const myComponent = renderer.create(
    <CounterContainer name="matt">
        <span>hello world</span>
        <span>hello world 2</span>
    </CounterContainer>).toJSON();
    expect(myComponent).toMatchSnapshot();
});
