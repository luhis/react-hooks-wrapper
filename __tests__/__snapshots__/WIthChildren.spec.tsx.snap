// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithChildren should 1`] = `
<div>
  <p>
    Hi 
    matt
    , You clicked 
    1
     times
  </p>
  <button
    onClick={[Function]}
  >
    Click me
  </button>
  <span>
    hello world
  </span>
  <span>
    hello world 2
  </span>
</div>
`;
